plutus-core/plutus-ir/src/PlutusIR/Transform/Inline.hs

```diff
@@ -74,12 +79,26 @@ contextually, so there's no point doing this.
 -- See Note [Inlining approach and 'Secrets of the GHC Inliner']
 newtype InlineTerm tyname name uni fun a = Done (Term tyname name uni fun a)
 
-newtype TermEnv tyname name uni fun a = TermEnv { unTermEnv :: UniqueMap TermUnique (InlineTerm tyname name uni fun a) }
+newtype TermEnv tyname name uni fun a = TermEnv { _unTermEnv :: UniqueMap TermUnique (InlineTerm tyname name uni fun a) }
     deriving newtype (Semigroup, Monoid)
 
-newtype Subst tyname name uni fun a = Subst { sTermEnv :: TermEnv tyname name uni fun a }
+newtype TypeEnv tyname uni a = TypeEnv { _unTypeEnv :: UniqueMap TypeUnique (Type tyname uni a) }
     deriving newtype (Semigroup, Monoid)
 
+data Subst tyname name uni fun a = Subst { _termEnv :: TermEnv tyname name uni fun a
+                                         , _typeEnv :: TypeEnv tyname uni a
+                                         }
+
+makeLenses ''TermEnv
+makeLenses ''TypeEnv
+makeLenses ''Subst
+
+instance Semigroup (Subst tyname name uni fun a) where
```

Comment:
If you like, you can use `Data.Semigroup.Generic` to derive this via `GenericSemigroup`, which is nice for records. Up to you, this is also fine.